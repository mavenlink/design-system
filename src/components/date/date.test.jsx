import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Date from './date.jsx';

describe('src/components/date/date.test.jsx', () => {
  const requiredProps = {
    id: 'test-id',
    label: 'Test label',
  };

  it('has defaults', () => {
    render(<Date {...requiredProps} />);
    expect(document.body).toMatchSnapshot();
  });

  it('validates on when the component loses focus', () => {
    render(<Date {...requiredProps} required={true} />);
    expect(screen.getByLabelText('Test label')).toBeRequired();
    expect(screen.getByText('(Required)')).toBeInTheDocument();
    expect(screen.getByLabelText('Test label')).toBeInvalid();
    expect(screen.getByLabelText('Test label')).toHaveDescription('');
    userEvent.click(screen.getByLabelText('Test label'));
    userEvent.tab({ shift: true });
    expect(screen.getByLabelText('Test label')).toBeInvalid();
    expect(screen.getByLabelText('Test label')).toHaveDescription('Constraints not satisfied');
  });

  it('updates the calendar when the user types', async () => {
    const localizedDate = (new window.Date('1999-01-01')).toLocaleDateString(undefined, {
      year: 'numeric',
      month: 'long',
    });

    render(<Date {...requiredProps} />);
    userEvent.type(screen.getByLabelText('Test label'), '1999-01-01');
    expect(await screen.findByText(localizedDate)).toBeInTheDocument();
  });

  describe('calendar behavior', () => {
    it('toggles open/close', () => {
      const today = (new window.Date()).toLocaleDateString(undefined, {
        year: 'numeric',
        month: 'long',
      });

      render(<Date {...requiredProps} />);
      userEvent.click(screen.getByTitle('Test label calendar button'));
      expect(screen.getByText(today)).toBeInTheDocument();
      userEvent.click(screen.getByTitle('Test label calendar button'));
      expect(screen.queryByText(today)).not.toBeInTheDocument();
    });

    it('closes on selection', () => {
      const today = (new window.Date()).toLocaleDateString(undefined, { month: 'long', day: 'numeric' });
      const todayText = (new window.Date()).toLocaleDateString(undefined, { month: 'short', year: 'numeric', day: 'numeric' });

      render(<Date {...requiredProps} />);
      userEvent.click(screen.getByTitle('Test label calendar button'));
      userEvent.click(screen.getByLabelText(today));
      expect(screen.queryByText(today)).not.toBeInTheDocument();
      expect(screen.getByLabelText('Test label')).toHaveValue(todayText);
    });

    it('closes on blur', () => {
      const today = (new window.Date()).toLocaleDateString(undefined, {
        year: 'numeric',
        month: 'long',
      });

      render(<Date {...requiredProps} />);
      userEvent.click(screen.getByTitle('Test label calendar button'));
      expect(screen.getByText(today)).toBeInTheDocument();
      userEvent.click(document.body);
      expect(screen.queryByText(today)).not.toBeInTheDocument();
    });

    it('closes on escape', () => {
      const today = (new window.Date()).toLocaleDateString(undefined, {
        year: 'numeric',
        month: 'long',
      });

      render(<Date {...requiredProps} />);
      userEvent.type(screen.getByLabelText('Test label'), '');
      expect(screen.getByText(today)).toBeInTheDocument();
      userEvent.type(screen.getByLabelText('Test label'), '{esc}', { skipClick: true });
      expect(screen.queryByText(today)).not.toBeInTheDocument();
    });
  });

  describe('input behavior', () => {
    it('is a text input', () => {
      render(<Date {...requiredProps} />);
      expect(screen.getByLabelText('Test label')).toHaveAttribute('type', 'text');
    });

    it('is a date input', () => {
      render(<Date {...requiredProps} />);
      userEvent.click(screen.getByLabelText('Test label'));
      expect(screen.getByLabelText('Test label')).toHaveAttribute('type', 'date');
    });
  });

  describe('id API', () => {
    it('is a string', () => {
      render(<Date {...requiredProps} id="unique-id" />);
      expect(screen.getByLabelText('Test label')).toHaveAttribute('id', 'unique-id');
    });
  });

  describe('label API', () => {
    it('is a string', () => {
      render(<Date {...requiredProps} label="Unique label" />);
      expect(screen.getByLabelText('Unique label')).toBeInTheDocument();
    });
  });

  describe('max API', () => {
    it('is date-only', () => {
      render(<Date {...requiredProps} max="2000-01-01" />);
      userEvent.type(screen.getByLabelText('Test label'), '2000-01-01');
      expect(screen.getByLabelText('Test label')).toBeValid();
      userEvent.clear(screen.getByLabelText('Test label'));
      userEvent.type(screen.getByLabelText('Test label'), '2000-01-02');
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('');
      userEvent.click(document.body);
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('Constraints not satisfied');
    });
  });

  describe('min API', () => {
    it('is date-only', () => {
      render(<Date {...requiredProps} min="2000-01-01" />);
      userEvent.type(screen.getByLabelText('Test label'), '2000-01-01');
      expect(screen.getByLabelText('Test label')).toBeValid();
      userEvent.clear(screen.getByLabelText('Test label'));
      userEvent.type(screen.getByLabelText('Test label'), '1999-12-31');
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('');
      userEvent.click(document.body);
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('Constraints not satisfied');
    });
  });

  describe('readOnly API', () => {
    it('is true', () => {
      const today = (new window.Date()).toLocaleDateString(undefined, {
        year: 'numeric',
        month: 'long',
      });

      render(<Date {...requiredProps} readOnly={true} />);
      expect(screen.getByLabelText('Test label')).toHaveAttribute('readonly');

      userEvent.click(screen.getByLabelText('Test label'));
      expect(screen.queryByText(today)).not.toBeInTheDocument();

      userEvent.type(screen.getByLabelText('Test label'), '{enter}');
      expect(screen.queryByText(today)).not.toBeInTheDocument();
    });

    it('is false', () => {
      const today = (new window.Date()).toLocaleDateString(undefined, {
        year: 'numeric',
        month: 'long',
      });

      render(<Date {...requiredProps} readOnly={false} />);
      expect(screen.getByLabelText('Test label')).not.toHaveAttribute('readonly');

      userEvent.tab();
      userEvent.type(screen.getByLabelText('Test label'), '{enter}', { skipClick: true });
      expect(screen.getByText(today)).toBeInTheDocument();
    });
  });

  describe('required API', () => {
    it('is true', () => {
      render(<Date {...requiredProps} required={true} />);
      expect(screen.getByLabelText('Test label')).toBeRequired();
      expect(screen.getByText('(Required)')).toBeInTheDocument();
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('');
      userEvent.click(screen.getByLabelText('Test label'));
      userEvent.click(document.body);
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('Constraints not satisfied');
    });

    it('is false', () => {
      render(<Date {...requiredProps} required={false} />);
      expect(screen.getByLabelText('Test label')).not.toBeRequired();
      expect(screen.queryByText('(Required)')).not.toBeInTheDocument();
    });
  });

  describe('validationMessage API', () => {
    it('is a string', () => {
      const { rerender } = render(<Date {...requiredProps} validationMessage="This is a provided error." />);
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('This is a provided error.');
      expect(screen.getByTitle('This is a provided error.')).toBeInTheDocument();

      rerender(<Date {...requiredProps} validationMessage="This is a new provided error." />);
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('This is a new provided error.');
      expect(screen.getByTitle('This is a new provided error.')).toBeInTheDocument();
    });

    it('is cleared', () => {
      const { rerender } = render(<Date {...requiredProps} validationMessage="This is a provided error." />);
      expect(screen.getByLabelText('Test label')).toBeInvalid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('This is a provided error.');
      expect(screen.getByTitle('This is a provided error.')).toBeInTheDocument();

      rerender(<Date {...requiredProps} validationMessage="" />);
      expect(screen.getByLabelText('Test label')).toBeValid();
      expect(screen.getByLabelText('Test label')).toHaveDescription('');
    });
  });
});
