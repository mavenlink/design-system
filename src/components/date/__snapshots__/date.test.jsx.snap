// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/date/date.test.jsx classNames API sets various class names 1`] = `
<body>
  <div>
    <div
      class="unique-container"
    >
      <div
        role="presentation"
      >
        <label
          class="label"
          for="test-id"
          id="test-id-label"
        >
          Test label
        </label>
        <div
          class="control-container"
        >
          <input
            aria-describedby="test-idHint"
            class="unique-input"
            id="test-id"
            name="test_name"
            type="text"
            value=""
          />
          <div
            class="icons"
          >
            <svg
              class="button"
              height="14"
              role="button"
              tabindex="0"
              width="11"
            >
              <title>
                Test label calendar button
              </title>
              <use
                xlink:href="#calendar.svg"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/date/date.test.jsx classNames API sets various class names 2`] = `
<body>
  <div>
    <div
      class="unique-container"
    >
      <div
        role="presentation"
      >
        <label
          class="label"
          for="test-id"
          id="test-id-label"
        >
          Test label
        </label>
        <div
          class="control-container"
        >
          <input
            aria-describedby="test-idHint"
            class="unique-input"
            id="test-id"
            name="test_name"
            type="date"
            value=""
          />
          <div
            class="icons"
          >
            <svg
              class="button"
              height="14"
              role="button"
              tabindex="0"
              width="11"
            >
              <title>
                Test label calendar button
              </title>
              <use
                xlink:href="#calendar.svg"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="unique-calendar-container"
      >
        <div
          class="container"
        >
          <div
            class="header"
          >
            <svg
              class="change-month-button"
              height="8"
              role="button"
              tabindex="0"
              width="12"
            >
              <title>
                Change calendar to March 2021
              </title>
              <use
                xlink:href="#arrow-left.svg"
              />
            </svg>
            <button
              class="year-button"
            >
              April 2021
              <svg
                class="change-year-icon"
                height="6"
                role="img"
                width="10"
              >
                <title>
                  Open year listbox
                </title>
                <use
                  xlink:href="#caret-down.svg"
                />
              </svg>
            </button>
            <svg
              class="change-month-button"
              height="8"
              role="button"
              tabindex="0"
              width="12"
            >
              <title>
                Change calendar to May 2021
              </title>
              <use
                xlink:href="#arrow-right.svg"
              />
            </svg>
          </div>
          <table
            class="calendar-grid"
            role="grid"
          >
            <thead>
              <tr>
                <th
                  class="weekday"
                >
                  S
                </th>
                <th
                  class="weekday"
                >
                  M
                </th>
                <th
                  class="weekday"
                >
                  T
                </th>
                <th
                  class="weekday"
                >
                  W
                </th>
                <th
                  class="weekday"
                >
                  T
                </th>
                <th
                  class="weekday"
                >
                  F
                </th>
                <th
                  class="weekday"
                >
                  S
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  aria-label="March 28"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1616889600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  28
                </td>
                <td
                  aria-label="March 29"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1616976000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  29
                </td>
                <td
                  aria-label="March 30"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1617062400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  30
                </td>
                <td
                  aria-label="March 31"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1617148800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  31
                </td>
                <td
                  aria-label="April 1"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617235200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  1
                </td>
                <td
                  aria-label="April 2"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617321600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  2
                </td>
                <td
                  aria-label="April 3"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617408000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  3
                </td>
              </tr>
              <tr>
                <td
                  aria-label="April 4"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617494400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  4
                </td>
                <td
                  aria-label="April 5"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617580800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  5
                </td>
                <td
                  aria-label="April 6"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617667200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  6
                </td>
                <td
                  aria-label="April 7"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617753600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  7
                </td>
                <td
                  aria-label="April 8"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617840000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  8
                </td>
                <td
                  aria-label="April 9"
                  aria-selected="false"
                  class="date"
                  data-epoch="1617926400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  9
                </td>
                <td
                  aria-label="April 10"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618012800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  10
                </td>
              </tr>
              <tr>
                <td
                  aria-label="April 11"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618099200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  11
                </td>
                <td
                  aria-label="April 12"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618185600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  12
                </td>
                <td
                  aria-label="April 13"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618272000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  13
                </td>
                <td
                  aria-label="April 14"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618358400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  14
                </td>
                <td
                  aria-label="April 15"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618444800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  15
                </td>
                <td
                  aria-label="April 16"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618531200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  16
                </td>
                <td
                  aria-label="April 17"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618617600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  17
                </td>
              </tr>
              <tr>
                <td
                  aria-label="April 18"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618704000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  18
                </td>
                <td
                  aria-label="April 19"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618790400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  19
                </td>
                <td
                  aria-label="April 20"
                  aria-selected="false"
                  class="highlighted-date"
                  data-epoch="1618876800000"
                  role="gridcell"
                  tabindex="0"
                >
                  20
                </td>
                <td
                  aria-label="April 21"
                  aria-selected="false"
                  class="date"
                  data-epoch="1618963200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  21
                </td>
                <td
                  aria-label="April 22"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619049600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  22
                </td>
                <td
                  aria-label="April 23"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619136000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  23
                </td>
                <td
                  aria-label="April 24"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619222400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  24
                </td>
              </tr>
              <tr>
                <td
                  aria-label="April 25"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619308800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  25
                </td>
                <td
                  aria-label="April 26"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619395200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  26
                </td>
                <td
                  aria-label="April 27"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619481600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  27
                </td>
                <td
                  aria-label="April 28"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619568000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  28
                </td>
                <td
                  aria-label="April 29"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619654400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  29
                </td>
                <td
                  aria-label="April 30"
                  aria-selected="false"
                  class="date"
                  data-epoch="1619740800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  30
                </td>
                <td
                  aria-label="May 1"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1619827200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  1
                </td>
              </tr>
              <tr>
                <td
                  aria-label="May 2"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1619913600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  2
                </td>
                <td
                  aria-label="May 3"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1620000000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  3
                </td>
                <td
                  aria-label="May 4"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1620086400000"
                  role="gridcell"
                  tabindex="-1"
                >
                  4
                </td>
                <td
                  aria-label="May 5"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1620172800000"
                  role="gridcell"
                  tabindex="-1"
                >
                  5
                </td>
                <td
                  aria-label="May 6"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1620259200000"
                  role="gridcell"
                  tabindex="-1"
                >
                  6
                </td>
                <td
                  aria-label="May 7"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1620345600000"
                  role="gridcell"
                  tabindex="-1"
                >
                  7
                </td>
                <td
                  aria-label="May 8"
                  aria-selected="false"
                  class="not-current-date"
                  data-epoch="1620432000000"
                  role="gridcell"
                  tabindex="-1"
                >
                  8
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/date/date.test.jsx has defaults 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        role="presentation"
      >
        <label
          class="label"
          for="test-id"
          id="test-id-label"
        >
          Test label
        </label>
        <div
          class="control-container"
        >
          <input
            aria-describedby="test-idHint"
            class="input"
            id="test-id"
            name="test_name"
            type="text"
            value=""
          />
          <div
            class="icons"
          >
            <svg
              class="button"
              height="14"
              role="button"
              tabindex="0"
              width="11"
            >
              <title>
                Test label calendar button
              </title>
              <use
                xlink:href="#calendar.svg"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/date/date.test.jsx has defaults 2`] = `
Object {
  "dirty": false,
  "id": "test-id",
  "name": "test_name",
  "value": undefined,
}
`;
